corehq/apps/enterprise/enterprise.py

```diff
@@ -635,3 +634,148 @@ def rows_for_domain(self, domain_obj):
+                'reporting_metadata.last_builds',
+            ]))
+        for user in user_query.run().hits:
+            last_builds = user.get('reporting_metadata', {}).get('last_builds', [])
+            for build in last_builds:
+                app_id = build.get('app_id')
+                build_version = build.get('build_version')
+                if app_id not in apps or not build_version:
+                    continue
+                build_version_date = DateTimeProperty.deserialize(build.get('build_version_date'))
+                all_builds = self.get_app_builds(domain, app_id, limit=self.INITIAL_QUERY_LIMIT)
+                latest_version = self.get_latest_build_version_at_time(all_builds, build_version_date)
+                if not latest_version:
+                    all_builds = self.get_app_builds(domain, app_id)
+                    latest_version = self.get_latest_build_version_at_time(all_builds, build_version_date)
+                yield user, build, latest_version
+
+    def get_app_builds(self, domain, app_id, limit=None):
+        if app_id not in self.builds_by_app_id:
+            app_es = (
+                AppES()
+                .domain(domain)
+                .is_build()
+                .app_id(app_id)
+                .sort('version', desc=True)
+                .is_released()
+                .source(['_id', 'version', 'last_released', 'built_on'])
+            )
+            if limit:
+                app_es = app_es.size(limit)
+            self.builds_by_app_id[app_id] = app_es.run().hits
+
+        return self.builds_by_app_id[app_id]
+
+    def get_latest_build_version_at_time(self, all_builds, time):
+        """
+        Get the latest build version at the time
+
+        :param all_builds: List of raw build documents sorted by version in descending order
+        :param time: The date of the build version to compare against
+        :return: The latest build version available at the given date
+        """
+
+        for build_doc in all_builds:
+            build_id = build_doc['_id']
+            if build_id in self.build_by_build_id:
+                build_info = self.build_by_build_id[build_id]
+            else:
+                # last_released is added in 2019, build before 2019 don't have this field
+                # TODO: have a migration to populate last_released from built_on
+                # Then this code can be modified to use last_released only
+                released_date = build_doc['last_released'] or build_doc['built_on']
+                build_info = {
+                    'version': build_doc['version'],
+                    'last_released': DateTimeProperty.deserialize(released_date)
+                }
+                self.build_by_build_id[build_id] = build_info
+
+            if build_info['last_released'] <= time:
+                return build_info['version']
+        return None
+
+
+def _format_percentage_for_enterprise_tile(dividend, divisor):
```

Comment:
Is it correct to have this marked as "private" with the leading `_`? It seems like it's used in a few different places in this file now.